### ISSUE TITLE
prune on a v2 repo requires multiple repack passes to remove all no longer needed stuff?

### ISSUE THREAD

**Issue by borkd** [restic/restic#3809]
Output of `restic version`
--------------------------
restic 0.13.1-dev (compiled manually) compiled with go1.18 on linux/amd64

How did you run restic exactly?
-------------------------------
restic-master prune --max-unused=0 --no-cache

- Using `--max-unused=0` now rather than leaving out for the default of 5%(?) after last couple of days or multiple passes, each taking many hours and not yielding expected results.
- `--no-cache` because repo is already on fast locally attached storage

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Prune is done locally on the machine holding the repo.
Backups were a mix of mostly local filesystem access, and some sftp.

Expected behavior
-----------------

Just like with v1 repos, prune needs a single pass to reduce v2 repo size to repack and remove all of the duplicate or otherwise unneeded data, taking however long such a pass takes.

Actual behavior
---------------

I have been trying to clean up a repo from 1.3TB down to the desired and expected 676.5GB. Almost 500GB+ were deemed duplicate by first run of `prune -vv`. Single prune operation takes considerable amount of time and removes maybe up to half of the the data it says it will remove. Looking at IO restic does a full pass over the repo in question, but does not remove all it say it would. Doing pass number 5 or 6 at the moment. Having the repo locked for this long is quite difficult.

```
[.. iteration 6? ..]
$ time ./restic-master prune --max-unused=0 --no-cache  ; echo EC=$?
repository 55b1eb2a opened (repo version 2) successfully, password is correct
warning: running prune without a cache, this may be very slow!
loading indexes...
loading all snapshots...
finding data that is still in use for 58 snapshots
[12:25] 100.00%  58 / 58 snapshots
searching used packs...
collecting packs for deletion and repacking
[0:04] 100.00%  158312 / 158312 packs processed

to repack:        537612 blobs / 156.912 GiB
this removes:      81536 blobs / 59.134 GiB
to delete:             0 blobs / 0 B
total prune:       81536 blobs / 59.134 GiB
remaining:      10299528 blobs / 676.441 GiB
unused size after prune: 0 B (0.00% of remaining size)

repacking packs
[2:58:15] 100.00%  33357 / 33357 packs repacked
rebuilding index
[0:17] 100.00%  156064 / 156064 packs processed
deleting obsolete index files
[0:00] 100.00%  69 / 69 files deleted
removing 33357 old packs
[2:45] 100.00%  33357 / 33357 files deleted
done

real    194m31.482s
user    157m30.135s
sys     8m15.705s
EC=0

$ du -shxc $RESTIC_REPOSITORY
728G    rrepo-master-v2/
728G    total

$ time ./restic-master prune --max-unused=0 --no-cache  ; echo EC=$?
repository 55b1eb2a opened (repo version 2) successfully, password is correct
warning: running prune without a cache, this may be very slow!
loading indexes...
loading all snapshots...
finding data that is still in use for 58 snapshots
[12:26] 100.00%  58 / 58 snapshots
searching used packs...
collecting packs for deletion and repacking
[0:04] 100.00%  156064 / 156064 packs processed

to repack:        463242 blobs / 134.960 GiB
this removes:      60106 blobs / 48.369 GiB
to delete:             0 blobs / 0 B
total prune:       60106 blobs / 48.369 GiB
remaining:      10299528 blobs / 676.441 GiB
unused size after prune: 0 B (0.00% of remaining size)

repacking packs
[2:36:25] 100.00%  28584 / 28584 packs repacked
rebuilding index
[0:17] 100.00%  154667 / 154667 packs processed
deleting obsolete index files
[0:00] 100.00%  69 / 69 files deleted
removing 28584 old packs
[2:20] 100.00%  28584 / 28584 files deleted
done

real    172m14.825s
user    143m13.570s
sys     8m1.939s
EC=0

```

Steps to reproduce the behavior
-------------------------------

- build restic from master branch
- initate v2 repo
- run a bunch of reasonably large backups against it, interrupt/disrupt enough of them to have sizable percentage of duplicate data
- run prune

Do you have any idea what may have caused this?
-----------------------------------------------

Pure speculation. I do not recall having such problems with same size or larger v1 repos, just with compressed v2 repos. Pruning duplicate data used to be a weak spot in the past (https://github.com/restic/restic/issues?q=is%3Aissue+label%3A%22category%3A+prune%22+) and some attempts were made to speed things up (https://github.com/restic/restic/pull/3290 or https://github.com/restic/restic/issues/2162). 

Having to recompress significant portion of the repo certainly does not help, but this issue is not purely about speedup but an issue in the existing algorithm which seems to require multiple passes to succeed. 

Do you have an idea how to solve the issue?
-------------------------------------------
I figured https://github.com/restic/restic/pull/3290 might be useful to speed up the operation, even if multiple passes would still be required, but it no longer cleanly applies on top of master branch

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Yes, and not just today

**Comment by aawsome** [comment]
I just realized that there is #3786 which has been merged and removed the needed pre-computation of duplicates blobs which is used in #3290. 
From a first glance, I'm not sure if there are side effects with #3786 and your behavior is such. Pinging @greatroar :wink: 

But if duplicates are the cause here, then there must be something wrong. `prune` is supposed to not leave any duplicates (at least without specifying a `max-repack-size` value).

As you are already using beta code and are on a local backend, can you compare with what [rusic](https://github.com/rustic-rs/rustic) is doing in your case? `rustic prune` include the logic of #3290 and other prune improvements. Running `restic prune` and `rustic-prune` with `--dry-run` should suffice.

**Comment by greatroar** [comment]
I don't see anything in the diff for #3786 that can cause this, but to be sure, @borkd should revert it and see if the version before that change. Either

```
git revert 8bdfcf779fb4e7260fc05649beb7c524d7518bbe
```

or, if that doesn't work for whatever reason,

```
git checkout 8bdfcf779fb4e7260fc05649beb7c524d7518bbe^  # Note the ^
```

then rebuild and retry. If that changes the behavior, we should revert #3786.

**Comment by aawsome** [comment]
I just a some time too take a look at #3786. IMO https://github.com/restic/restic/blob/master/cmd/restic/cmd_prune.go#L376 is the problem.

With #3786, duplicates are no longer computed a priori, but in the same iteration where for each pack the decision what to do is made. Now, let's say we have pack A and pack B with identical blobs.
When A gets processed first, these blobs are marked as used. If all blobs in A are either used or duplicates, then by the code above, A is marked for keeping. Even worse: When B gets processed after A, it is marked for repacking. BUT: The Repack functions works globally, i.e. it gets all packs to repack and all blobs to keep and then keeps these blobs only once when repacking. Now, in our example, A is not repacked, but all blobs which exist in A are also kept in B. So the repacking does not even remove all duplicates from B.

@borkd: You should see this difference cleary when running the current `restic prune --dry-run` against the reverted version. Or against `rustic prune --dry-run`

**Comment by borkd** [comment]
@greatroar - I have applied 3810 and restic is building.

@aawsome - rustic (rustic-rs-v0.2.0-x86_64-unknown-linux-musl) appears to be confused, output is below:

```
$ RUST_BACKTRACE=full ./rustic-rs -r rrepo-master-v2/ prune --dry-run
enter repository password:
password is correct
reading index...
[00:00:01] ████████████████████████████████████████         69/69                                                                                                                                                             reading snapshots...
[00:00:00] ████████████████████████████████████████         58/58                                                                                                                                                             finding used blobs...
[00:00:05] ████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░          7/58                                                                                                                                                             thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: unknown variant `dev`, expected one of `file`, `dir`, `symlink`, `device` at line 1 column 323', src/blob/tree.rs:203:33
stack backtrace:
   0:     0x772cda656c9d - <unknown>
   1:     0x772cda69530c - <unknown>
   2:     0x772cda650ab1 - <unknown>
   3:     0x772cda6587c5 - <unknown>
   4:     0x772cda658479 - <unknown>
   5:     0x772cda658d12 - <unknown>
   6:     0x772cda658bf7 - <unknown>
   7:     0x772cda657154 - <unknown>
   8:     0x772cda658929 - <unknown>
   9:     0x772cda0ad343 - <unknown>
  10:     0x772cda0ad433 - <unknown>
  11:     0x772cda120837 - <unknown>
  12:     0x772cda118a4a - <unknown>
  13:     0x772cda0e816d - <unknown>
  14:     0x772cda1d174d - <unknown>
  15:     0x772cda131b5a - <unknown>
  16:     0x772cda14516d - <unknown>
  17:     0x772cda1742fb - <unknown>
  18:     0x772cda163066 - <unknown>
  19:     0x772cda120a13 - <unknown>
  20:     0x772cda136b2d - <unknown>
  21:     0x772cda64a236 - <unknown>
  22:     0x772cda163142 - <unknown>
```

**Comment by borkd** [comment]
@greatroar and @aawsome - seems #3810 does the trick.

```
./restic-master-3810-linux-amd64 prune --max-unused=0 --no-cache  ; echo EC=$?
repository 55b1eb2a opened (repo version 2) successfully, password is correct
warning: running prune without a cache, this may be very slow!
loading indexes...
loading all snapshots...
finding data that is still in use for 58 snapshots
[10:52] 100.00%  58 / 58 snapshots
searching used packs...
collecting packs for deletion and repacking
[0:04] 100.00%  154667 / 154667 packs processed

to repack:        754656 blobs / 220.011 GiB
this removes:      48226 blobs / 41.395 GiB
to delete:             0 blobs / 0 B
total prune:       48226 blobs / 41.395 GiB
remaining:      10299528 blobs / 676.441 GiB
unused size after prune: 0 B (0.00% of remaining size)

repacking packs
[3:53:48] 100.00%  46607 / 46607 packs repacked
rebuilding index
[0:17] 100.00%  146438 / 146438 packs processed
deleting obsolete index files
[0:00] 100.00%  69 / 69 files deleted
removing 46607 old packs
[4:18] 100.00%  46607 / 46607 files deleted
done

EC=0

$ du -shxc rrepo-master-v2/
680G    rrepo-master-v2/
680G    total
```
